File: IPaCoreDataController/IPaCoreDataController.m
//
//  IPaCoreDataController.m
//
//  Created by IPaPa on 2011/6/8.
//  Copyright 2011 __MyCompanyName__. All rights reserved.
//

#import "IPaCoreDataController.h"


@implementation IPaCoreDataController
{
	NSString *DBName;
}
#pragma mark - life cycle

- (IPaCoreDataController*) initDBWithName:(NSString*)dbName withParam:(NSDictionary*)params
{
    self = [super init];
	DBName = [dbName stringByAppendingString:@".sqlite"];
    
    
    //initial managed object model
    id value = params[IPaCoreDataPKey_SingleObjectModel];
    if (value) {
        NSString *modelPath = [[NSBundle mainBundle] pathForResource:value ofType:@"momd"];
        
        if (!modelPath) {
            modelPath = [[NSBundle mainBundle] pathForResource:value ofType:@"mom"];
        }
        
        
        // If path is nil, then NSURL or NSManagedObjectModel will throw an exception
        
        NSURL *momUrl = [NSURL fileURLWithPath:modelPath];
        _managedObjectModel = [[NSManagedObjectModel alloc] initWithContentsOfURL:momUrl];
    }
    else {
        _managedObjectModel = [NSManagedObjectModel mergedModelFromBundles:nil];
    }

    //initial persistentStoreCoordinator
    value = params[IPaCoreDataPKey_DBPath];
	NSString *storePath = (value)?value:NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES)[0];
    
    storePath = [storePath stringByAppendingPathComponent:DBName];
	/*
	 Set up the store.
	 For the sake of illustration, provide a pre-populated default store.
	 */
	NSURL *storeUrl = [NSURL fileURLWithPath:storePath];
	NSError *error;
    _persistentStoreCoordinator = [[NSPersistentStoreCoordinator alloc] initWithManagedObjectModel: [self managedObjectModel]];
    if (![_persistentStoreCoordinator addPersistentStoreWithType:NSSQLiteStoreType configuration:nil URL:storeUrl options:nil error:&error]) {
		/*
		 Replace this implementation with code to handle the error appropriately.
		 
		 abort() causes the application to generate a crash log and terminate. You should not use this function in a shipping application, although it may be useful during development. If it is not possible to recover from the error, display an alert panel that instructs the user to quit the application by pressing the Home button.
		 
		 Typical reasons for an error here include:
		 * The persistent store is not accessible
		 * The schema for the persistent store is incompatible with current managed object model
		 Check the error message to determine what the actual problem was.
		 */
		NSLog(@"Unresolved error %@, %@", error, [error userInfo]);
		abort();
    }   
	
	
	//initial managedObjectContext

    if (_persistentStoreCoordinator != nil) {
        _managedObjectContext = [NSManagedObjectContext new];
        [_managedObjectContext setPersistentStoreCoordinator: _persistentStoreCoordinator];
    }


	return self;
}



#pragma mark - core data stack

- (void)save
{
	NSError *error = nil;
	NSManagedObjectContext *context = self.managedObjectContext;
    if (context != nil)
    {
        if ([context hasChanges] && ![context save:&error])
        {
            /*
             Replace this implementation with code to handle the error appropriately.
             
             abort() causes the application to generate a crash log and terminate. You should not use this function in a shipping application, although it may be useful during development. If it is not possible to recover from the error, display an alert panel that instructs the user to quit the application by pressing the Home button.
             */
            NSLog(@"Unresolved error %@, %@", error, [error userInfo]);
            abort();
        } 
    }
}

- (id) insertNewObjectForEntityForName:(NSString*)entityName;
{
	return [NSEntityDescription insertNewObjectForEntityForName:entityName inManagedObjectContext:self.managedObjectContext];
	
}

- (void) deleteObject:(NSManagedObject *)object
{
	[_managedObjectContext deleteObject:object];
	[self save];
}
- (NSFetchedResultsController*)getFetchedResultsWithEntity:(NSString*)entityName 
												   SortKey:(NSString*)sortKey 
												 Ascending:(bool)ascending 
												 CacheName:(NSString*)cacheName
{	
	// Create the fetch request for the entity.
	NSFetchRequest *fetchRequest = [[NSFetchRequest alloc] init];
	// Edit the entity name as appropriate.
	NSEntityDescription *entity = [NSEntityDescription entityForName:entityName inManagedObjectContext:_managedObjectContext];
	[fetchRequest setEntity:entity];
	// Edit the sort key as appropriate.
	NSSortDescriptor *sortDescriptor = [[NSSortDescriptor alloc] initWithKey:sortKey ascending:ascending];
	NSArray *sortDescriptors = @[sortDescriptor];
	[fetchRequest setSortDescriptors:sortDescriptors];        
	// Edit the section name key path and cache name if appropriate.
	// nil for section name key path means "no sections".
	NSFetchedResultsController *aFetchedResultsController = [[NSFetchedResultsController alloc] initWithFetchRequest:fetchRequest managedObjectContext:_managedObjectContext sectionNameKeyPath:nil cacheName:cacheName];
	
	//NSLog(@"%d",[aFetchedResultsController retainCount]);
	//return [aFetchedResultsController autorelease];
    return aFetchedResultsController;
}
-(NSArray*)fetchResultWithEntityName:(NSString*)entityName
                           Predicate:(NSPredicate*)predicate
                      SortDescriptors:(NSArray*)SortDescriptors
                          FetchLimit:(NSUInteger)FetchLimit
{

    NSManagedObjectContext *moc = [self managedObjectContext];
    
    NSEntityDescription *entity = [NSEntityDescription entityForName:entityName
                                              inManagedObjectContext:moc];
    
    NSFetchRequest *fetchRequest = [[NSFetchRequest alloc] init];
    [fetchRequest setEntity:entity];
    [fetchRequest setPredicate:predicate];
    [fetchRequest setSortDescriptors:SortDescriptors];
    [fetchRequest setFetchLimit:FetchLimit];
    
    return [moc executeFetchRequest:fetchRequest error:nil];
    
}

@end

File: IPaCoreDataController/IPaCoreDataController.swift
//
//  IPaCoreDataController.swift
//  IPaCoreDataController
//
//  Created by IPa Chen on 2015/8/16.
//  Copyright (c) 2015年 A Magic Studio. All rights reserved.
//

import Foundation
import CoreData
public class IPaCoreDataController :NSObject{
    
    var managedObjectModel:NSManagedObjectModel
    lazy var persistentStoreCoordinator:NSPersistentStoreCoordinator = {
        let persistentStoreCoordinator = NSPersistentStoreCoordinator(managedObjectModel: self.managedObjectModel)
        do {
            try persistentStoreCoordinator.addPersistentStoreWithType(self.sourceStoreType, configuration: nil, URL: self.dbStoreURL, options: [NSMigratePersistentStoresAutomaticallyOption: true,
                NSInferMappingModelAutomaticallyOption: true])
            /*
            Replace this implementation with code to handle the error appropriately.
            
            abort() causes the application to generate a crash log and terminate. You should not use this function in a shipping application, although it may be useful during development. If it is not possible to recover from the error, display an alert panel that instructs the user to quit the application by pressing the Home button.
            
            Typical reasons for an error here include:
            * The persistent store is not accessible
            * The schema for the persistent store is incompatible with current managed object model
            Check the error message to determine what the actual problem was.
            */
            
            
            
        }
        catch let error as NSError {
            print("Unresolved error \(error), \(error.userInfo)")
            abort();
            
        }
        return persistentStoreCoordinator
    }()
    lazy var managedObjectContext:NSManagedObjectContext = {
        let managedObjectContext = NSManagedObjectContext(concurrencyType: .MainQueueConcurrencyType)
        managedObjectContext.persistentStoreCoordinator = self.persistentStoreCoordinator
        return managedObjectContext
    }()
    var sourceStoreType:String {
        get {
            return NSSQLiteStoreType
        }
    }
    var dbStoreURL:NSURL
    init(dbName:String,dbPath:String?,modelName:String?) {

        
        let dbFileName = (dbName as NSString).stringByAppendingPathExtension("sqlite")!
        var momdUrl:NSURL?


        if let modelName = modelName {
            var modelPath:String? = nil
            modelPath = NSBundle.mainBundle().pathForResource(modelName, ofType: "momd")
            if modelPath == nil {
                modelPath = NSBundle.mainBundle().pathForResource(modelName, ofType: "mom")
            }
            momdUrl = NSURL(fileURLWithPath: modelPath!)

        }
        if let momdUrl = momdUrl {
            managedObjectModel = NSManagedObjectModel(contentsOfURL: momdUrl)!

        }
        else {
            managedObjectModel = NSManagedObjectModel.mergedModelFromBundles(nil)!

        }
        var storePath:String
        if let dbPath = dbPath {
            storePath = dbPath
        }
        else {
            storePath = (NSSearchPathForDirectoriesInDomains(.DocumentDirectory,.UserDomainMask, true).first)!
        }

        //bug handling
        
        let bugFileName = (dbName as NSString).stringByAppendingPathExtension(".sqlite")!
        let fileManager = NSFileManager.defaultManager()
        let bugFilePath = (storePath as NSString).stringByAppendingPathComponent(bugFileName)
        
        storePath = (storePath as NSString).stringByAppendingPathComponent(dbFileName)
        
        
        if fileManager.fileExistsAtPath(bugFilePath) {
            do {
                try fileManager.moveItemAtPath(bugFilePath, toPath: storePath)
            }
            catch {
                
            }
        }
        
        
        
        
        dbStoreURL = NSURL(fileURLWithPath: storePath)
        

        /*
        Set up the store.
        For the sake of illustration, provide a pre-populated default store.
        */
       
    }
    convenience init(dbName:String,modelName:String)
    {
        self.init(dbName:dbName,dbPath:nil,modelName:modelName)
    }
    convenience init(modelName:String)
    {
        self.init(dbName:modelName,modelName:modelName)
    }
    //MARK: Migration

    public func makeMigration() -> Bool {
        let sourceURL = dbStoreURL
        let bundle = NSBundle.mainBundle()
        var sourceModel:NSManagedObjectModel
        do {
            let sourceMetadata = try NSPersistentStoreCoordinator.metadataForPersistentStoreOfType(sourceStoreType, URL: sourceURL, options: nil)
            if managedObjectModel.isConfiguration(nil, compatibleWithStoreMetadata: sourceMetadata) {
                return true
            }
            sourceModel = NSManagedObjectModel.mergedModelFromBundles([bundle], forStoreMetadata: sourceMetadata)!
            
        }
        catch let error as NSError {
            print("\(error)")
            return false
        }
        catch {
            return false
        }
        //search model paths in bundle
        var targetModel:NSManagedObjectModel?
        var targetMappingModel:NSMappingModel?
        var targetModelName:String?
        let momdArray = bundle.pathsForResourcesOfType("momd", inDirectory: nil)
        
        for momdPath in momdArray {
            
            let resourceSubpath = (momdPath as NSString).lastPathComponent
            let array = bundle.pathsForResourcesOfType("mom", inDirectory: resourceSubpath)
            for momPath in array {
                


                if let model =  NSManagedObjectModel(contentsOfURL: NSURL(fileURLWithPath: momPath)) {
                    
                    targetMappingModel = NSMappingModel(fromBundles:[bundle], forSourceModel: sourceModel, destinationModel: model)
                    if let _ = targetMappingModel {
                        targetModelName = ((momPath as NSString).lastPathComponent as NSString).stringByDeletingPathExtension
                        targetModel = model
                        break
                    }
                }
            }
            if let _ = targetModel  {
                break
            }
        }
        if targetModel == nil {
            let otherModels = bundle.pathsForResourcesOfType("mom", inDirectory: nil)
            for momPath in otherModels {
                if let model =  NSManagedObjectModel(contentsOfURL: NSURL(fileURLWithPath: momPath)) {
                    
                    
                    targetMappingModel = NSMappingModel(fromBundles:[bundle], forSourceModel: sourceModel, destinationModel: targetModel)
                    if let _ = targetMappingModel {
                        targetModelName = ((momPath as NSString).lastPathComponent as NSString).stringByDeletingPathExtension
                        targetModel = model
                        break
                    }
                }
            }
        }
        guard let destinationModel = targetModel else {
            print("destination model not found!!")
            return false
        }
        targetMappingModel = NSMappingModel(fromBundles:[bundle], forSourceModel: sourceModel, destinationModel: destinationModel)
        if targetMappingModel == nil {
            do {
                targetMappingModel = try NSMappingModel.inferredMappingModelForSourceModel(sourceModel, destinationModel: destinationModel)
            }
            catch let error as NSError {
                print("\(error)")
                return false
            }
            catch {
                return false
            }
        }
        guard let destinationMappingModel = targetMappingModel else {
            print("can not create Mapping Model")
            return false
        }
        // Build a path to write the new store
        let storePath:NSString = sourceURL.path!
        let destinationURL = NSURL(fileURLWithPath:"\(storePath.stringByDeletingPathExtension).\(targetModelName!).\(storePath.pathExtension)")
        //start migration
        let manager = NSMigrationManager(sourceModel: sourceModel, destinationModel: destinationModel)
        manager.addObserver(self, forKeyPath: "migrationProgress", options: .New, context: nil)
        do {
            try manager.migrateStoreFromURL(sourceURL, type: sourceStoreType, options: nil, withMappingModel: destinationMappingModel, toDestinationURL: destinationURL, destinationType: sourceStoreType, destinationOptions: nil)
            manager.removeObserver(self, forKeyPath: "migrationProgress")
        }
        catch let error as NSError {
            print("\(error)")
            manager.removeObserver(self, forKeyPath: "migrationProgress")
            return false
        }
        catch {
            manager.removeObserver(self, forKeyPath: "migrationProgress")
            return false
        }

        // Migration was successful, move the files around to preserve the source in case things go bad
        let guid = NSProcessInfo.processInfo().globallyUniqueString
        let backupPath = (NSTemporaryDirectory() as NSString).stringByAppendingPathComponent(guid)
        let fileManager = NSFileManager.defaultManager()
        do {
            try fileManager.moveItemAtPath(sourceURL.path!, toPath: backupPath)
            try fileManager.moveItemAtPath(destinationURL.path!, toPath: sourceURL.path!)
            
        }
        catch let error as NSError {
            print("\(error)")
            return false
        }
        catch {
            return false
        }

        return makeMigration()
        
    }
    //MARK: Public
    
    public func checkMigration() -> Bool {
        
        guard let path = self.dbStoreURL.path where NSFileManager.defaultManager().fileExistsAtPath(path) else {
            return false
        }
        //check migration
        do {

            let sourceMetadata = try NSPersistentStoreCoordinator.metadataForPersistentStoreOfType(self.sourceStoreType, URL: self.dbStoreURL, options: nil)

                // Migration is needed if destinationModel is NOT compatible
            if !managedObjectModel.isConfiguration(nil, compatibleWithStoreMetadata: sourceMetadata) {
                //migration needed

                
                
                
            }


        }
        catch let error as NSError {
            print("\(error)");
            return false
        }
        catch {
            
        }
        return true
        
        
    }
    public func deleteEntity(entityName:String) {
        let fetchAllObjects = NSFetchRequest()
        fetchAllObjects.entity = NSEntityDescription.entityForName(entityName, inManagedObjectContext: managedObjectContext)
        fetchAllObjects.includesPropertyValues = false //only fetch the managedObjectID

        do {
            let allObjects = try managedObjectContext.executeFetchRequest(fetchAllObjects)
            for object in allObjects {
                if let object = object as? NSManagedObject {
                    managedObjectContext.deleteObject(object)
                }
            }
        } catch let error as NSError {

            print("\(error)");
        }

        

    }
    //MARK: core data stack
    public func save() {
        do {
            if managedObjectContext.hasChanges {
                try managedObjectContext.save()
            }
        } catch let error as NSError {
            
            print("Unresolved error \(error), \(error.userInfo)");
           
            abort();
        }
    }
    
    
    public func insertNewObject(entityName:String) -> NSManagedObject {
        return NSEntityDescription.insertNewObjectForEntityForName(entityName, inManagedObjectContext: managedObjectContext) 
    }
    public func deleteObject(object:NSManagedObject) {
        managedObjectContext.deleteObject(object)
        save()
    }
    public func fetch(request:NSFetchRequest) -> [AnyObject]? {
        var fetchResult:[AnyObject]?
        do {
            try fetchResult = managedObjectContext.executeFetchRequest(request)
        } catch let error as NSError {
            
            print("Unresolved error \(error), \(error.userInfo)");
            
            abort();
        }
        return fetchResult
    }
    public func createWorkerManagedContex(concurencyType:NSManagedObjectContextConcurrencyType) -> NSManagedObjectContext {
        let workerMOC = NSManagedObjectContext(concurrencyType: concurencyType)
        workerMOC.parentContext = managedObjectContext
        return workerMOC
    }
    
    //MARK:Observer
    public override func observeValueForKeyPath(keyPath: String?, ofObject object: AnyObject?, change: [String : AnyObject]?, context: UnsafeMutablePointer<Void>) {
        if keyPath == "migrationProgress" ,let manager = object as? NSMigrationManager {
            print("progress \(manager.migrationProgress)")
        }
        else {
            super.observeValueForKeyPath(keyPath, ofObject: object, change: change, context: context)
        }
    }
}


